### PR TITLE
[Blazor] Cleans up infrastructure used for tests

### PR DESCRIPTION
**Pull request by javiercn**
Simplifies the test infrastructure to avoid relying on a cookie.

### CODE REVIEW

**Comment by javiercn** [comment]
Ping

File: src/Components/test/testassets/BasicTestApp/wwwroot/index.html [hunk 1]
```diff
@@ -25,22 +25,8 @@
         function navigationManagerNavigate() {
             Blazor.navigateTo('/subdir/some-path');
         }
-
-        (function () {
-            // Load either blazor.webassembly.js or blazor.server.js depending
-            // on the hash part of the URL. This is just to give a way for the
-            // test runner to make the selection.
-            if (location.hash === '#server') {
```
Comment (SteveSandersonMS) [inline src/Components/test/testassets/BasicTestApp/wwwroot/index.html]: What's the updated way to launch from VS with server-side execution?

File: src/Components/Components.sln [hunk 2]
```diff
@@ -236,6 +232,12 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.AspNetCore.Compon
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.AspNetCore.Components.Forms.Tests", "Forms\test\Microsoft.AspNetCore.Components.Forms.Tests.csproj", "{173D84A3-0F37-480F-AC0F-7E2DBBE32B28}"
 EndProject
+Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Samples", "Samples", "{46E4300C-5726-4108-B9A2-18BB94EB26ED}"
+EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "BlazorServer.Sample", "Samples\BlazorServer.Sample\BlazorServer.Sample.csproj", "{2D7B741B-FD32-4BA6-9C98-CCC5B424B005}"
```
Comment (SteveSandersonMS) [inline src/Components/Components.sln]: Any reason for having `.Sample` in the name? It's already in a directory called `Samples`. Could this be something like `BlazorServerApp`? That would be more inline with the naming patterns for the other apps.

File: src/Components/Components.sln [hunk 3]
```diff
@@ -236,6 +232,12 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.AspNetCore.Compon
...
+EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.AspNetCore.HttpsPolicy", "..\Middleware\HttpsPolicy\src\Microsoft.AspNetCore.HttpsPolicy.csproj", "{4664276D-606A-4BB3-873A-9EE84FB22877}"
+EndProject
```
Comment (SteveSandersonMS) [inline src/Components/Components.sln]: Why add the `HttpsPolicy` project to this sln?

File: src/Components/test/E2ETest/ServerExecutionTests/PrerenderingTest.cs [hunk 4]
```diff
@@ -119,5 +121,8 @@ private void BeginInteractivity()
         {
             Browser.FindElement(By.Id("load-boot-script")).Click();
         }
+
+        private void SignInAs(string usernName, string roles, bool useSeparateTab = false) =>
```
Comment (SteveSandersonMS) [inline src/Components/test/E2ETest/ServerExecutionTests/PrerenderingTest.cs]: Typo: `usernName`

File: src/Components/test/E2ETest/ServerExecutionTests/GlobalizationTest.cs [hunk 5]
```diff
@@ -8,29 +8,30 @@
...
+            : base(browserFixture, serverFixture, output)
         {
+            serverFixture.BuildWebHostMethod = serverFixture.BuildWebHostMethod = TestServer.Program.BuildWebHost<InternationalizationStartup>;
         }
 
         protected override void InitializeAsyncCore()
         {
             // On WebAssembly, page reloads are expensive so skip if possible
-            Navigate(ServerPathBase, _serverFixture.ExecutionMode == ExecutionMode.Client);
+            Navigate(ServerPathBase);
```
Comment (SteveSandersonMS) [inline src/Components/test/E2ETest/ServerExecutionTests/GlobalizationTest.cs]: This line of codes no longer seems to correspond with the comment above it. Can you update one of them so they match?

Also does this mean we're now always doing a full reload on WebAssembly mode? Does that make the test run substantially longer?

File: src/Components/test/E2ETest/ServerExecutionTests/LocalizationTest.cs [hunk 6]
```diff
@@ -7,27 +7,29 @@
...
             ITestOutputHelper output)
-            : base(browserFixture, serverFixture.WithServerExecution(), output)
+            : base(browserFixture, serverFixture, output)
         {
+            serverFixture.BuildWebHostMethod = TestServer.Program.BuildWebHost<InternationalizationStartup>;
         }
 
         protected override void InitializeAsyncCore()
         {
             // On WebAssembly, page reloads are expensive so skip if possible
```
Comment (SteveSandersonMS) [inline src/Components/test/E2ETest/ServerExecutionTests/LocalizationTest.cs]: Outdated comment?

File: src/Components/test/testassets/TestServer/Startup.cs [hunk 7]
```diff
@@ -32,6 +32,13 @@ public void ConfigureServices(IServiceCollection services)
                     var detailedErrors = Configuration.GetValue<bool>("circuit-detailed-errors");
                     o.DetailedErrors = detailedErrors;
                 });
+            services.AddAuthentication(CookieAuthenticationDefaults.AuthenticationScheme).AddCookie();
+
+            services.AddAuthorization(options =>
+            {
+                options.AddPolicy("NameMustStartWithB", policy =>
+                    policy.RequireAssertion(ctx => ctx.User.Identity.Name?.StartsWith("B") ?? false));
+            });
```
Comment (SteveSandersonMS) [inline src/Components/test/testassets/TestServer/Startup.cs]: Why is this auth stuff going in here? Shouldn't it only be in the auth-specific startup class?

File: src/Components/test/testassets/TestServer/.gitignore [hunk 8]
```diff
@@ -0,0 +1 @@
+wwwroot/**/*
```
Comment (SteveSandersonMS) [inline src/Components/test/testassets/TestServer/.gitignore]: What's the reason for doing all the copying on build, rather than serving the contents from `../BasicTestApp/wwwroot`?

File: src/Components/test/E2ETest/Infrastructure/BasicTestAppTestBase.cs [hunk 9]
```diff
@@ -1,25 +0,0 @@
-// Copyright (c) .NET Foundation. All rights reserved.
```
Comment (SteveSandersonMS) [inline src/Components/test/E2ETest/Infrastructure/BasicTestAppTestBase.cs]: Removing this makes a bunch of other tests seem more complicated (because they have to inherit from `ServerTestBase<ToggleExecutionModeServerFixture<Program>>`). What's the advantage of removing it?

File: src/Components/test/E2ETest/Infrastructure/WebDriverExtensions/BasicTestAppAuthenticationWebDriverExtensions.cs [hunk 10]
```diff
@@ -4,7 +4,7 @@
 
 namespace Microsoft.AspNetCore.Components.E2ETest
 {
-    internal static class AuthHelpers
+    internal static class BasicTestAppAuthenticationWebDriverExtensions
```
Comment (SteveSandersonMS) [inline src/Components/test/E2ETest/Infrastructure/WebDriverExtensions/BasicTestAppAuthenticationWebDriverExtensions.cs]: ~~Helpers~~

I'm sure @rynowak will be pleased :)

File: src/Components/test/testassets/TestServer/ClientStartup.cs [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+
+namespace TestServer
+{
+    // WARNING: DO NOT MODIFY THIS STARTUP CLASS FOR TEST PURPOSES
+    // Most of the client-side tests are executed using the developer host directly, so changing values here won't
+    // affect any client-side test.
```
Comment (SteveSandersonMS) [inline src/Components/test/testassets/TestServer/ClientStartup.cs]: It seems a bit odd to have it then.

Would it be possible to instead invoke the dev server code to host the client-side tests? It should be possible to do that from `Program.cs` in the same way that the tests themselves do.

File: src/Components/test/testassets/TestServer/Startup.cs [hunk 12]
```diff
@@ -31,41 +31,47 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
...
+        table, th, td, tr { border: 1px solid black; }
+        th { font-weight: bold; }
+    <style>
+  </body>
+</html>");
+                var content = writer.ToString();
+                ctx.Response.ContentLength = content.Length;
+                using var responseWriter = new StreamWriter(ctx.Response.Body);
+                await responseWriter.WriteAsync(content);
+                await responseWriter.FlushAsync();
```
Comment (SteveSandersonMS) [inline src/Components/test/testassets/TestServer/Startup.cs]: I don't really feel strongly, but it might even be worth putting in a Razor page to render this content :)

File: src/Components/test/testassets/TestServer/Startup.cs (hunk 7 above)
Comment (SteveSandersonMS) [reply src/Components/test/testassets/TestServer/Startup.cs]: Looks like you also removed it in a later commit. Correct?

File: src/Components/test/testassets/BasicTestApp/wwwroot/index.html (hunk 1 above)
Comment (javiercn) [reply src/Components/test/testassets/BasicTestApp/wwwroot/index.html]: You just F5 and you will be presented with an html page with links to each scenario that has a custom startup

File: src/Components/Components.sln (hunk 3 above)
Comment (javiercn) [reply src/Components/Components.sln]: The sample app needs it and I want to keep https enabled in the sample to reflect the template as close as possible

File: src/Components/test/E2ETest/ServerExecutionTests/GlobalizationTest.cs (hunk 5 above)
Comment (javiercn) [reply src/Components/test/E2ETest/ServerExecutionTests/GlobalizationTest.cs]: Yeah, this was a bad copy pasta in the tests. These tests only run on blazor server-side

File: src/Components/test/testassets/TestServer/Startup.cs (hunk 7 above)
Comment (javiercn) [reply src/Components/test/testassets/TestServer/Startup.cs]: I took it out I believe

File: src/Components/test/E2ETest/Infrastructure/BasicTestAppTestBase.cs (hunk 9 above)
Comment (javiercn) [reply src/Components/test/E2ETest/Infrastructure/BasicTestAppTestBase.cs]: That it gets miss used. We were inheriting from BasicTestAppTestBase on a bunch of places (and getting a `ToggleExecutionModeServerFixture<Program>` in a bunch of places for server-only tests and that made it hard to determine what tests run where.

This way is one less concept and you are forced to choose the right fixture, which makes it clear what the scenario for the test is.

File: src/Components/test/E2ETest/Infrastructure/WebDriverExtensions/BasicTestAppAuthenticationWebDriverExtensions.cs (hunk 10 above)
Comment (javiercn) [reply src/Components/test/E2ETest/Infrastructure/WebDriverExtensions/BasicTestAppAuthenticationWebDriverExtensions.cs]: Yeah, git is doing weird stuff. This class/file is new

File: src/Components/test/testassets/BasicTestApp/wwwroot/index.html (hunk 1 above)
Comment (SteveSandersonMS) [reply src/Components/test/testassets/BasicTestApp/wwwroot/index.html]: OK, that sounds great. Thanks!

File: src/Components/test/testassets/TestServer/ClientStartup.cs (hunk 11 above)
Comment (javiercn) [reply src/Components/test/testassets/TestServer/ClientStartup.cs]: Its fair. I though of doing it that way, but one thing that I like about having a separate startup here is that you can modify it while you are troubleshooting stuff and everything is contained within the same project.

File: src/Components/Components.sln (hunk 3 above)
Comment (SteveSandersonMS) [reply src/Components/Components.sln]: Ah yes. I've been working from the `.slnf` for so long I forgot the `.sln` has to include all the shared-framework-type dependencies.

File: src/Components/test/E2ETest/Infrastructure/BasicTestAppTestBase.cs (hunk 9 above)
Comment (SteveSandersonMS) [reply src/Components/test/E2ETest/Infrastructure/BasicTestAppTestBase.cs]: OK that's fine. Thanks for explaining.

File: src/Components/test/testassets/TestServer/Startup.cs (hunk 12 above)
Comment (javiercn) [reply src/Components/test/testassets/TestServer/Startup.cs]: I tried to be as minimalist as possible here to avoid any unwanted interference that I can't foresee. I don't expect this bit to change so I'm not concerned.

File: src/Components/test/E2ETest/Infrastructure/WebDriverExtensions/BasicTestAppWebDriverExtensions.cs [hunk 13]
```diff
@@ -0,0 +1,24 @@
+using OpenQA.Selenium;
+using OpenQA.Selenium.Support.UI;
+
+namespace Microsoft.AspNetCore.Components.E2ETest
+{
+    internal static class BasicTestAppWebDriverExtensions
+    {
+        public static IWebElement MountTestComponent<TComponent>(this IWebDriver browser) where TComponent : IComponent
```
Comment (rynowak) [inline src/Components/test/E2ETest/Infrastructure/WebDriverExtensions/BasicTestAppWebDriverExtensions.cs]: Do we need this as both an extension method and protected method on a base class?

File: src/Components/test/E2ETest/Infrastructure/WebDriverExtensions/BrowserExtensions.cs [hunk 14]
```diff
@@ -0,0 +1,58 @@
+using System;
+using System.Collections.Generic;
+using OpenQA.Selenium;
+using OpenQA.Selenium.Support.UI;
+using Xunit.Sdk;
+
+namespace Microsoft.AspNetCore.Components.E2ETest
+{
+    internal static class BrowserExtensions
```
Comment (rynowak) [inline src/Components/test/E2ETest/Infrastructure/WebDriverExtensions/BrowserExtensions.cs]: nit: `WebDriverExtensions`

File: src/Components/test/E2ETest/Tests/AuthTest.cs [hunk 15]
```diff
@@ -12,7 +13,7 @@
 namespace Microsoft.AspNetCore.Components.E2ETest.Tests
 {
     [Collection("auth")] // Because auth uses cookies, this can't run in parallel with other auth tests
-    public class AuthTest : BasicTestAppTestBase
```
Comment (rynowak) [inline src/Components/test/E2ETest/Tests/AuthTest.cs]: I expected to see more files moving around since we were going to organize tests by which startup they use? Not part of this PR?

File: src/Components/test/E2ETest/Tests/HttpClientTest.cs [hunk 16]
```diff
@@ -1,43 +1,50 @@
...
-using System.Threading.Tasks;
+using TestServer;
 using Xunit;
 using Xunit.Abstractions;
 
 namespace Microsoft.AspNetCore.Components.E2ETest.Tests
 {
-    public class HttpClientTest : BasicTestAppTestBase, IClassFixture<AspNetSiteServerFixture>
+    public class HttpClientTest : BrowserTestBase,
+        IClassFixture<DevHostServerFixture<BasicTestApp.Program>>, IClassFixture<BasicTestAppServerSiteFixture<CorsStartup>>
```
Comment (rynowak) [inline src/Components/test/E2ETest/Tests/HttpClientTest.cs]: Should we introduce another base-class here? Noticing that this doesn't use `ServerTestBase` and had to duplicate the log clearing logic.

File: src/Components/test/E2ETest/Tests/HttpClientTest.cs [hunk 17]
```diff
@@ -47,14 +54,14 @@ public async Task SanityCheck_ApiServerIsRunning()
             // Note that the HttpClient we're instantiating here is *not* the
             // one under test. This is not related to Blazor in any way.
             var httpClient = new HttpClient { BaseAddress = _apiServerFixture.RootUri };
-            var responseText = await httpClient.GetStringAsync("/api/greeting/sayhello");
+            var responseText = await httpClient.GetStringAsync("/subdir/api/greeting/sayhello");
```
Comment (rynowak) [inline src/Components/test/E2ETest/Tests/HttpClientTest.cs]: meta question - directed at @SteveSandersonMS as well as @javiercn - does it make sense that we're making base-paths a part of literally all of these tests?

I would expect us to have a few tests for base paths: something hello-worldey as well as some routing tests. It looks like it's **everywhere**

File: src/Components/test/testassets/TestServer/.gitignore (hunk 8 above)
Comment (rynowak) [reply src/Components/test/testassets/TestServer/.gitignore]: +1 for wanting to know the answer.

File: src/Components/test/testassets/TestServer/Program.cs [hunk 18]
```diff
@@ -11,9 +16,33 @@ public class Program
...
+                ["CORS (WASM)"] = (BuildWebHost<CorsStartup>(CreateAdditionalArgs(args)), "/subdir"),
+                ["Prerendering (Server-side)"] = (BuildWebHost<PrerenderedStartup>(CreateAdditionalArgs(args)), "/prerendered"),
+                ["Globalization + Localization (Server-side)"] = (BuildWebHost<InternationalizationStartup>(CreateAdditionalArgs(args)), "/subdir"),
+                ["Server-side blazor"] = (BuildWebHost<ServerStartup>(CreateAdditionalArgs(args)), "/subdir"),
+                ["Hosted client-side blazor"] = (BuildWebHost<ClientStartup>(CreateAdditionalArgs(args)), "/subdir")
+            };
+
+            var mainHost = BuildWebHost(args);
+
+            Task.WhenAll(createIndividualHosts.Select(s => s.Value.host.StartAsync())).GetAwaiter().GetResult();
```
Comment (rynowak) [inline src/Components/test/testassets/TestServer/Program.cs]: super - S P I C Y

File: src/Components/test/E2ETest/Tests/HttpClientTest.cs (hunk 17 above)
Comment (SteveSandersonMS) [reply src/Components/test/E2ETest/Tests/HttpClientTest.cs]: Since nonempty base paths are the sort of subtle thing that could trip us up when implementing any feature that involves URLs, I think it's good that we involve them by default.

I'd certainly be happy to also have a special-case test showing that an empty base path works too.

Comment (rynowak) [reply src/Components/test/E2ETest/Tests/HttpClientTest.cs]: So routing and navigation tests then? I don't get why this is this is a concern that cross-cuts all tests.

File: src/Components/Components.sln (hunk 2 above)
Comment (javiercn) [reply src/Components/Components.sln]: Renamed to BlazorServerApp

File: src/Components/test/testassets/TestServer/.gitignore (hunk 8 above)
Comment (javiercn) [reply src/Components/test/testassets/TestServer/.gitignore]: The files in BasicTestApp/wwwroot were being served by `UseBlazorClientSideStaticfiles` files. ServerStartup does not add `UseBlazorClientSideStaticFiles`.

Some tests for blazor server-side and blazor client-side use these files. I didn't want to copy the files as that might get them out of date, so instead I chose to copy them at build time and added the .gitignore file so that they don't show up as modified on every build change.

I could choose to remove the .gitignore and add them to source control. But we might run into issues where they get out of date. Id rather keep the current solution.

File: src/Components/test/testassets/TestServer/ClientStartup.cs (hunk 11 above)
Comment (javiercn) [reply src/Components/test/testassets/TestServer/ClientStartup.cs]: Added an option to start from the devhost server.

File: src/Components/test/E2ETest/Tests/AuthTest.cs (hunk 15 above)
Comment (javiercn) [reply src/Components/test/E2ETest/Tests/AuthTest.cs]: Not part of this PR. I'm trying to this in steps and this is already a pretty big PR

File: src/Components/test/E2ETest/Tests/HttpClientTest.cs (hunk 16 above)
Comment (javiercn) [reply src/Components/test/E2ETest/Tests/HttpClientTest.cs]: Switched it to use ServerTestBase

File: src/Components/test/E2ETest/Tests/HttpClientTest.cs (hunk 17 above)
Comment (javiercn) [reply src/Components/test/E2ETest/Tests/HttpClientTest.cs]: I'm fine with whatever way we choose to go here, but I'm not tackling that as part of this PR.

File: src/Components/test/testassets/TestServer/Program.cs [hunk 19]
```diff
@@ -1,19 +1,90 @@
...
+            {
+                "--contentroot", contentRoot,
+                "--pathbase", "/subdir",
+                "--applicationpath", typeof(BasicTestApp.Program).Assembly.Location,
+            }).ToArray();
+            var host = DevServerProgram.BuildWebHost(finalArgs);
+            return (new WebHostShim(host), "/subdir");
         }
 
+        private class WebHostShim : IWebHost
```
Comment (javiercn) [inline src/Components/test/testassets/TestServer/Program.cs]: This is temporary, I'm planning to send out a separate PR where I convert all the test infrastructure to use IHost. I just didn't want to increase the scope of this PR even more.

File: src/Components/test/E2ETest/Infrastructure/WebDriverExtensions/BasicTestAppWebDriverExtensions.cs (hunk 13 above)
Comment (javiercn) [reply src/Components/test/E2ETest/Infrastructure/WebDriverExtensions/BasicTestAppWebDriverExtensions.cs]: I took the base class methods out. WaitUntilTestSelectorReady and WaitUntilExists. I've combined them in the WaitAssert class so that they all respect the same timeouts that the other infrastructure improvements PR has.

**Comment by javiercn** [comment]
🆙📅

**Comment by javiercn** [comment]
There are a few cleanups left that I'm leaving out from this PR:
* Move to IHost instead of IWebHost in fixture classes.
* Fully standarize asserts
  * Got rid of some asserts on the base class, but I'm planning to do more cleanup/standarization.
  * There are some Asserts (WaitUntilExists) that need to be rationalized.
* Organize test components based on scenario:
  * I didn't want to make this PR even bigger.
  * I think we can alternatively organize them by TestClass instead of by Startup class, that would provide
    better distribution and make test code easier to find based on the test class.

**Comment by javiercn** [comment]
This has integrated the changes for Selenium infrastructure improvements now. Below is a sample exception with fast fail on errors in the browser log

image

There were a bunch of things using WaitUntilExists without checking the logs for fast failure. I've switched them to use Browser.Exists so that they respect the new configurable default timeouts.

**Comment by javiercn** [comment]
Anything else here? I don't want the PR to stale.

File: src/Components/test/E2ETest/Infrastructure/WebDriverExtensions/BasicTestAppWebDriverExtensions.cs [hunk 20]
```diff
@@ -17,7 +18,7 @@ internal static class BasicTestAppWebDriverExtensions
         public static SelectElement WaitUntilTestSelectorReady(this IWebDriver browser)
         {
             var elemToFind = By.CssSelector("#test-selector > select");
-            browser.WaitUntilExists(elemToFind, timeoutSeconds: 30, throwOnError: true);
+            browser.Exists(elemToFind);
```
Comment (SteveSandersonMS) [inline src/Components/test/E2ETest/Infrastructure/WebDriverExtensions/BasicTestAppWebDriverExtensions.cs]: I don't follow the difference between `browser.Exists` and `browser.WaitUntilExists`, considering that both of their implementations involve waiting.

Do we need to have both of these methods? Could we drop `Exists`, since the name is harder to understand?

File: src/Shared/E2ETesting/WaitAssert.cs [hunk 21]
```diff
@@ -50,10 +50,13 @@ public static void Exists(this IWebDriver driver, By finder)
         public static void Exists(this IWebDriver driver, By finder, TimeSpan timeout)
             => WaitAssertCore(driver, () => Assert.NotEmpty(driver.FindElements(finder)), timeout);
 
-        public static IWebElement WaitUntilExists(this IWebDriver driver, By findBy) =>
-            WaitUntilExists(driver, findBy, default);
+        public static IWebElement Element(this IWebDriver driver, By findBy)
+        {
+            driver.Exists(findBy);
+            return driver.FindElements(findBy)[0];
```
Comment (SteveSandersonMS) [inline src/Shared/E2ETesting/WaitAssert.cs]: The naming of these `IWebElement` extension methods bothers me a bit, because nothing in their names makes clear which of them tolerate waiting until the thing exists or is true, and which of them will return null or fail of the thing/condition isn't initially true.

It's not enough just to say "we should always wait" because (1) the native Selenium methods don't, and nothing stops us using them, and (2) it's not always desirable to wait. For the tests to be as precise as possible, in scenarios where we know something should already be true, it's better to use a lookup/assertion that would fail immediately if it couldn't be satisfied on the first attempt.

How would you feel about changing our "wait until ... " extension methods to have names that all start with "Wait"? I have thought this for a while about methods like `Browser.AssertX`, which I always thought should be called `Browser.WaitAssertX` for clarity.

If you would be happy with this change but don't want to do it in this PR, that's totally fine with me. If we have agreement we can create a follow-up issue for it, which I don't mind doing if you're busy with other things.

File: src/Components/test/testassets/TestServer/.gitignore (hunk 8 above)
Comment (SteveSandersonMS) [reply src/Components/test/testassets/TestServer/.gitignore]: I was definitely not suggesting we should make static copies of the files in source control. I agree that would get out-of-date very quickly.

I was wondering whether we had the option to serve the files at runtime from `../BasicTestApp/wwwroot`. That would avoid the copying *and* prevents any of us in the future from getting confused about why editing the copies under `TestServer` doesn't take effect.

Do we have that option? I think it would be greatly preferable if we do have it.

**Review by SteveSandersonMS — APPROVED** [review]
Looks great!

Hope you don't mind, but I added a couple more questions and [unresolved one of the conversations](https://github.com/aspnet/AspNetCore/pull/13515#pullrequestreview-282932288). If you're able to react to those before merging that would be great. Please IM me if you want to discuss!

File: src/Components/test/testassets/TestServer/.gitignore (hunk 8 above)
Comment (javiercn) [reply src/Components/test/testassets/TestServer/.gitignore]: @SteveSandersonMS We could add an extra file provider to do that I think, but I tried to stay away from that as it moved the complexity from build time to run time and I wanted to keep the services/middleware config as close as possible to our default apps.

My approach was weighted on the fact that we don't change these files or add new files often, which is why I think it strikes a good balance between keeping the runtime code simple and avoid duplicating/checking-in files.

Does that make sense?

File: src/Components/test/E2ETest/Infrastructure/WebDriverExtensions/BasicTestAppWebDriverExtensions.cs (hunk 20 above)
Comment (javiercn) [reply src/Components/test/E2ETest/Infrastructure/WebDriverExtensions/BasicTestAppWebDriverExtensions.cs]: This was a temporary change, I think I changed it to WaitUntilExists.

The main difference between WaitUntilExists and Exists is that the first one will throw an exception inmediately when observing errors in the BrowserLogs without waiting till the end. So this will fail faster than even a regular Assert.

Arguably we could get rid of WaitUntilExists in favor of Exists as the main value for the first one was development, which is addressed with the change that I did for allowing different configurations for dev/ci, and fail fast after first failure.

The other difference is that Exists doesn't return the found element and WaitUntilExists does, but we can fix that.

Would you be ok if we integrate everything under Exists?

File: src/Shared/E2ETesting/WaitAssert.cs (hunk 21 above)
Comment (javiercn) [reply src/Shared/E2ETesting/WaitAssert.cs]: Should we simply integrate all of them within Exists and have Exists return the element if found? I was trying to avoid making this PR into an "assertion consolidation PR", but I'm fine taking that change so long we are all ok with it.

To be clear the proposal would be something like

Exists `IWebElement Exists(this IWebDriver driver, Func<IWebElement> elementSelector, TimeSpan timeout = default)

Comment (SteveSandersonMS) [reply src/Shared/E2ETesting/WaitAssert.cs]: +1 to integrating them into a smaller number of methods.

My question about the naming still stands though. Do you have any objection to using the term `Wait` in the method names for all the things that implicitly wait?

Comment (javiercn) [reply src/Shared/E2ETesting/WaitAssert.cs]: * I've integrated everything under Browser.Exist (which now returns the found element).
* Every method in `WaitAssert.cs` implicitly waits for the assertion to be true or fails otherwise.
* This is inline with how `Assert.Single` and similar methods work on xUnit.

File: src/Components/test/E2ETest/Infrastructure/WebDriverExtensions/BasicTestAppWebDriverExtensions.cs (hunk 20 above)
Comment (javiercn) [reply src/Components/test/E2ETest/Infrastructure/WebDriverExtensions/BasicTestAppWebDriverExtensions.cs]: Everything here is integrated under Browser.Exists (which now returns the element)
